### PR TITLE
[INTEL MKL] Enable NCHW to NHWC conversion for CPU - part1.

### PR DESCRIPTION
**Pull request by mdfaijul**
This PR enables layout (data format) conversion from NCHW to NHWC on CPU. This is useful (1) when a model is trained on GPU and later doing inference/fine-tuning on CPU and (2) helps quantizing NCHW trained model for CPU.

To make the PR small, we have included a few unit tests in this **part1**. More unit tests will follow on future PRs.

### CODE REVIEW

File: tensorflow/core/grappler/optimizers/generic_layout_optimizer_test.cc [hunk 1]
```diff
@@ -49,6 +49,31 @@ constexpr int kDepthIn = 8;
 constexpr int kKernel = 3;
 constexpr int kDepthOut = 16;
 
+#if (GOOGLE_CUDA || TENSORFLOW_USE_ROCM)
+#define DIMS(n, h, w, c) \
```
Comment (ezhulenev) [inline tensorflow/core/grappler/optimizers/generic_layout_optimizer_test.cc:53]: Could you please add few lines of documentation explaining what's going on here, so people without a context can get what's up quickly. Something like:

"When GPU available we test .... when only CPU is available we ..."

**Review by ezhulenev — CHANGES_REQUESTED** [review]
Thanks for the change, we also need it for the tf-js.

**Comment by gbaned** [comment]
@mdfaijul Can you please check @ezhulenev's comments and keep us posted. Thanks!

File: tensorflow/core/grappler/optimizers/generic_layout_optimizer_test.cc (hunk 1 above)
Comment (gzmkl) [reply tensorflow/core/grappler/optimizers/generic_layout_optimizer_test.cc:53]: OK. I will follow-up as I am covering the developer who is on vacation now

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

**Comment by mdfaijul** [comment]
@ezhulenev Sorry, I have been in vacation. I will resume working this PR shortly.

**Comment by mdfaijul** [comment]
@ezhulenev Thanks for the suggestions. I have added comments. Please check.

**Comment by mdfaijul** [comment]
@ezhulenev Any update on this PR?

**Comment by mdfaijul** [comment]
@ezhulenev Any update on this PR?

**Comment by mdfaijul** [comment]
@ezhulenev Did you have any chance looking into changes to address your comments?

**Comment by mdfaijul** [comment]
@ezhulenev @gbaned Did you have chance to look into the updates in response to review comments.

**Review by ezhulenev — DISMISSED** [review]

**Comment by ezhulenev** [comment]
This PR failed internally:

```
tensorflow/core/grappler/optimizers:generic_layout_optimizer_test |  
Build Flags: --config=cuda --compilation_mode=opt
```

```
third_party/tensorflow/core/grappler/optimizers/generic_layout_optimizer_test.cc:224: Failure
Expected equality of these values:
  attr->s()
    Which is: "NHWC"
  attr_value
    Which is: "NCHW"
```

**Comment by gbaned** [comment]
@mdfaijul  Can you please check @ezhulenev's comments and keep us posted ? Thanks!

**Comment by mdfaijul** [comment]
@gbaned Thanks. I am looking into the error. Will get back to you shortly.

**Comment by mdfaijul** [comment]
@gbaned @ezhulenev When committed a fix to the GPU failure from another machine where git was not setup well, I got CLA issues. The PR #42448 is identical to this closed one except with small fix to the GPU failure. Sorry for the inconvenience.
